pkg/skaffold/config/portforward.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"strconv"
+)
+
+// These are the list of accepted port-forward modes.
+const (
+	// user enables user-defined port-forwards.
+	user = "user"
+	// services enables forwarding Kubernetes services.
+	services = "services"
+	// debug enables forwarding just debug-related containerPorts on pods.
+	debug = "debug"
+	// pods enables forwarding of all containerPorts on pods.
+	pods = "pods"
+	// off disables port forwarding.
+	off = "off"
+)
+
+// PortForwardOptions are options set by the command line for port forwarding.
+// `off` is intended to be a single standalone option.
+type PortForwardOptions struct {
+	Modes []string
+}
+
+// Enabled checks if the port-forwarding options indicates that forwarding should be enabled.
+func (p PortForwardOptions) Enabled() bool {
+	if len(p.Modes) == 0 {
+		return false
+	}
+	// --port-forward was previously a boolean, so accept pflag's boolean values.
+	// This method accepts "off" or bool variants to be mixed with others,
+	// leaving complaining to be done by Validate.
+	for _, o := range p.Modes {
+		b, err := strconv.ParseBool(o)
+		if err == nil && !b {
+			return false
+		}
+		if o == off {
+			return false
+		}
+	}
+	return true
```

Comment:
this block is duplicated several times in this file, with the only difference between them being the target string when you're searching through the modes. it might be cleaner to extract this away as a helper method - something like 

```golang
func (p PortForwardOptions) ModeEnabled(mode string) bool {
  for _, o := range p.Modes {
	  	b, err := strconv.ParseBool(o)
		  if err == nil && !b {
			  return false
		  }
		  if o == mode {
			  return false
		  }
	}
}
```

Reply:
The code snippets are subtly different.  But this code can be made table-driven which would be easier to reason about.

Reply:
So making it table driven made the problem worse, but I was able to simplify the functions.